flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -205,6 +206,8 @@
 
 	private Long syncSavepointId = null;
 
+	private final Map<StreamOperator<?>, ProcessingTimeServiceImpl> processingTimeServices;
```

Comment:
> map key: Can we use OperatorID? StreamOperator is public interface and we know nothing about it's hashcode/equals (and we don't hold extra references to operators then)

Reply:
> map value: It's better to specify an interface, not a concrete implementation to depend on it less.

The methods `quiesce` and `getTimersDoneFutureAfterQuiescing` need to be added in the processing time service, but the `ProcessingTimeService` interafce is user-oriented, so we don't want to change it due to the runtime needs. On the other hand, `StreamTask` understands `ProcessingTimeServiceImpl` because it uses `ProcessingTimeServiceImpl`  to build instances in `#getProcessingTimeService`. So it should be possible to use the concrete implementation here.

Reply:
Yes, it is possible but if we later will want to return another implementation or different implementations then it will have to change the signature.